overlord/snapstate/check_snap.go

```diff
@@ -99,40 +99,86 @@ func checkVersion(version string) bool {
 	return true
 }
 
-var openSnapFile = backend.OpenSnapFile
+type ErrSnapNeedsMode struct {
+	Mode snap.ConfinementType
+	Snap string
+}
 
-// checkSnap ensures that the snap can be installed.
-func checkSnap(st *state.State, snapFilePath string, si *snap.SideInfo, curInfo *snap.Info, flags Flags) error {
-	// This assumes that the snap was already verified or --dangerous was used.
+func (e *ErrSnapNeedsMode) Error() string {
+	return fmt.Sprintf("snap %q requires %s or confinement override", e.Snap, e.Mode)
+}
 
-	s, _, err := openSnapFile(snapFilePath, si)
-	if err != nil {
-		return err
-	}
+type ErrSnapNeedsClassicSystem string
 
-	if s.NeedsDevMode() && !flags.DevModeAllowed() {
-		return fmt.Errorf("snap %q requires devmode or confinement override", s.Name())
-	}
-	if s.NeedsClassic() {
-		if !release.OnClassic {
-			return fmt.Errorf("snap %q requires classic confinement which is only available on classic systems", s.Name())
+func (e ErrSnapNeedsClassicSystem) Error() string {
+	return fmt.Sprintf("snap %q requires classic confinement which is only available on classic systems", string(e))
+}
+
+func checkSnapInfoFlags(info *snap.Info, snapst *SnapState, flags Flags) error {
```

Comment:
this probably shouldn't have a name starting with checkSnap, is really checking the flags, not the the snap